src/library/scala/concurrent/impl/Promise.scala

```diff
@@ -298,40 +311,40 @@ private[concurrent] object Promise {
         if (state.isInstanceOf[Try[T]]) {
           if(!target.tryComplete0(target.get(), state.asInstanceOf[Try[T]]))
             throw new IllegalStateException("Cannot link completed promises together")
-        } else if (state.isInstanceOf[Link[T]]) {
-          state.asInstanceOf[Link[T]].promise(this).linkRootOf(target, link)
-        } else /*if (state.isInstanceOf[Callbacks[T]]) */ {
+        } else if (state.isInstanceOf[Callbacks[T]]) {
           val l = if (link ne null) link else new Link(target)
           val p = l.promise(this)
-          if (p ne this) {
-            if (compareAndSet(state, l)) {
-              if (state ne Noop) p.dispatchOrAddCallbacks(p.get(), state.asInstanceOf[Callbacks[T]]) // Noop-check is important here
-            } else linkRootOf(p, l)
-          }
-        }
+          if ((this ne p) && compareAndSet(state, l)) {
+            if (state ne Noop) p.dispatchOrAddCallbacks(p.get(), state.asInstanceOf[Callbacks[T]]) // Noop-check is important here
+          } else linkRootOf(p, l)
```

Comment:
@jrudolph Yes, the change is intentional. Since the method is tail recursive it will be a no-op if it recurses: https://github.com/scala/scala/pull/7663/files#diff-e0c06059ba197fea5927c4e4b0f3d837R309